### ISSUE TITLE
Unable To Parse Content 'NoneType' object has no attribute 'split'

### ISSUE THREAD

**Issue by DunaSpice** [neo4j-labs/llm-graph-builder#914]
I get this error when trying to graph a file. Docker and local deploys. Tried all I could. Sorry if this is just config issues but I spent a good day o this now.  in the front it shows Failed To Process File:Documentation for local deployments - Neo4j Labs.pdf or LLM Unable To Parse Content 'NoneType' object has no attribute 'split'

thanks.

2024-12-03 00:19:34,638 - Received notification from DBMS server: {severity: WARNING} {code: Neo.ClientNotification.Statement.FeatureDeprecationWarning} {category: DEPRECATION} {title: This feature is deprecated and will be removed in future versions.} {description: The query used a deprecated field from a procedure. ('advertisedListenAddress' returned by 'gds.debug.arrow' is deprecated.)} {position: line: 1, column: 1, offset: 0} for query: 'CALL gds.debug.arrow()'
2024-12-03 00:19:34,638 - Received notification from DBMS server: {severity: WARNING} {code: Neo.ClientNotification.Statement.FeatureDeprecationWarning} {category: DEPRECATION} {title: This feature is deprecated and will be removed in future versions.} {description: The query used a deprecated field from a procedure. ('serverLocation' returned by 'gds.debug.arrow' is deprecated.)} {position: line: 1, column: 1, offset: 0} for query: 'CALL gds.debug.arrow()'
2024-12-03 00:19:34,638 - Successfully created GDS driver.
2024-12-03 00:19:34,638 - Starting to clear communities.
2024-12-03 00:19:34,638 - Dropping communities...
2024-12-03 00:19:34,647 - Communities dropped successfully
2024-12-03 00:19:34,647 - Dropping community property from entities...
2024-12-03 00:19:34,654 - Community property dropped successfully
2024-12-03 00:19:34,668 - Received notification from DBMS server: {severity: WARNING} {code: Neo.ClientNotification.Statement.FeatureDeprecationWarning} {category: DEPRECATION} {title: This feature is deprecated and will be removed in future versions.} {description: The query used a deprecated field from a procedure. ('schema' returned by 'gds.graph.list' is deprecated.)} {position: line: 1, column: 1, offset: 0} for query: 'CALL gds.graph.list()'
2024-12-03 00:19:34,668 - Creating new graph project 'communities'.
2024-12-03 00:19:34,692 - Received notification from DBMS server: {severity: WARNING} {code: Neo.ClientNotification.Statement.UnknownLabelWarning} {category: UNRECOGNIZED} {title: The provided label is not in the database.} {description: One of the labels in your query is not available in the database, make sure you didn't misspell it or that the label is available when you run this statement in your application (the missing label name is: __Community__)} {position: line: 2, column: 33, offset: 33} for query: "\nMATCH (source:!Chunk&!Document&!__Community__)-[]->(target:!Chunk&!Document&!__Community__)\nWITH source, target, count(*) as weight\nWITH gds.graph.project(\n               'communities',\n               source,\n               target,\n               {\n               relationshipProperties: { weight: weight }\n               },\n               {undirectedRelationshipTypes: ['*']}\n               ) AS g\nRETURN\n  g.graphName AS graph_name, g.nodeCount AS nodes, g.relationshipCount AS rels\n"
2024-12-03 00:19:34,693 - Received notification from DBMS server: {severity: WARNING} {code: Neo.ClientNotification.Statement.UnknownLabelWarning} {category: UNRECOGNIZED} {title: The provided label is not in the database.} {description: One of the labels in your query is not available in the database, make sure you didn't misspell it or that the label is available when you run this statement in your application (the missing label name is: __Community__)} {position: line: 2, column: 78, offset: 78} for query: "\nMATCH (source:!Chunk&!Document&!__Community__)-[]->(target:!Chunk&!Document&!__Community__)\nWITH source, target, count(*) as weight\nWITH gds.graph.project(\n               'communities',\n               source,\n               target,\n               {\n               relationshipProperties: { weight: weight }\n               },\n               {undirectedRelationshipTypes: ['*']}\n               ) AS g\nRETURN\n  g.graphName AS graph_name, g.nodeCount AS nodes, g.relationshipCount AS rels\n"
2024-12-03 00:19:34,694 - Graph projection 'None' created successfully with None nodes and None relationships.
2024-12-03 00:19:34,704 - Failed to create community graph project: No projected graph named 'None' exists in current database 'neo4j'
2024-12-03 00:19:34,704 - Failed to create communities: No projected graph named 'None' exists in current database 'neo4j'
2024-12-03 00:19:34,704 - created communities
[DEFAULT]{'api_name': 'post_processing/create_communities', 'db_url': 'bolt://localhost:7687', 'logging_time': '2024-12-03 08:19:34 UTC'}
INFO:     127.0.0.1:50125 - "POST /post_processing HTTP/1.1" 200 OK

**Comment by kartikpersistent** [comment]
Hi @DunaSpice can you share few details like backend env config, which model you tried  ?

**Comment by kaustubh-darekar** [comment]
Hi @DunaSpice 
the actual error might be just above the logs you shared here. Can you share whole logs trail.

also check if your model is defined as per example.env (https://github.com/neo4j-labs/llm-graph-builder/blob/main/backend/example.env)
for example if you are using openai gpt 4o model you can define

LLM_MODEL_CONFIG_openai_gpt_4o="gpt-4o-2024-11-20,openai_api_key"


let us know if this solves the issue you are facing

**Comment by DunaSpice** [comment]
> Hi @DunaSpice can you share few details like backend env config, which model you tried ?

here is the .env with some parts of secret ____ removed. I have also tried to connect to AURA with same result.
OPENAI_API_KEY="________skmhx_5SxMSgO9KPD-9vB6p3oA"
DIFFBOT_API_KEY="________3faa049bfe6f977971fd7c742200"
GROQ_API_KEY="_______Y2Mmzs5KWjQmOiFne7OBkUNtS"
EMBEDDING_MODEL = "all-MiniLM-L6-v2"
RAGAS_EMBEDDING_MODEL = "openai"
IS_EMBEDDING = "true"
KNN_MIN_SCORE = "0.94"
# Enable Gemini (default is False) | Can be False or True
GEMINI_ENABLED = False
# Enable Google Cloud logs (default is False) | Can be False or True
GCP_LOG_METRICS_ENABLED = False
NUMBER_OF_CHUNKS_TO_COMBINE = 6
UPDATE_GRAPH_CHUNKS_PROCESSED = 20
NEO4J_URI = "bolt://host.docker.internal"
NEO4J_USERNAME = "neo4j"
NEO4J_PASSWORD = "_____"
NEO4J_DATABASE = "neo4j"
AWS_ACCESS_KEY_ID =  ""
AWS_SECRET_ACCESS_KEY = ""
LANGCHAIN_API_KEY = ""
LANGCHAIN_PROJECT = ""
LANGCHAIN_TRACING_V2 = ""
LANGCHAIN_ENDPOINT = ""
GCS_FILE_CACHE = "" #save the file into GCS or local, SHould be True or False
NEO4J_USER_AGENT=""
ENABLE_USER_AGENT = ""
LLM_MODEL_CONFIG_model_version=""
ENTITY_EMBEDDING="" True or False
DUPLICATE_SCORE_VALUE = ""
DUPLICATE_TEXT_DISTANCE = ""
#examples
LLM_MODEL_CONFIG_openai-gpt-3.5="gpt-3.5-turbo-0125,openai_api_key"
LLM_MODEL_CONFIG_openai_gpt_4o_mini="gpt-4o-mini-2024-07-18,openai_api_key"
LLM_MODEL_CONFIG_gemini_1.5_pro="gemini-1.5-pro-002"
LLM_MODEL_CONFIG_gemini_1.5_flash="gemini-1.5-flash-002"
LLM_MODEL_CONFIG_diffbot="diffbot,diffbot_api_key"
LLM_MODEL_CONFIG_azure_ai_gpt_35="azure_deployment_name,azure_endpoint or base_url,azure_api_key,api_version"
LLM_MODEL_CONFIG_azure_ai_gpt_4o="gpt-4o,https://YOUR-ENDPOINT.openai.azure.com/,azure_api_key,api_version"
LLM_MODEL_CONFIG_groq_llama3_70b="model_name,base_url,groq_api_key"
LLM_MODEL_CONFIG_anthropic_claude_3_5_sonnet="model_name,anthropic_api_key"
LLM_MODEL_CONFIG_fireworks_llama_v3_70b="model_name,fireworks_api_key"
LLM_MODEL_CONFIG_bedrock_claude_3_5_sonnet="model_name,aws_access_key_id,aws_secret__access_key,region_name"
LLM_MODEL_CONFIG_ollama_llama3="model_name,model_local_url"
YOUTUBE_TRANSCRIPT_PROXY="https://user:pass@domain:port"

**Comment by DunaSpice** [comment]
not sure but maybe this has to do with this (INFO:     Started reloader process [40271] using WatchFiles
/Users/peter/miniconda3/envs/my_env_name/lib/python3.10/site-packages/pydantic/_internal/_fields.py:161: UserWarning: Field "model_name" has conflict with protected namespace "model_".

You may be able to resolve this warning by setting `model_config['protected_namespaces'] = ()`.
  warnings.warn(
2024-12-03 02:44:07,415 - Python-dotenv could not parse statement starting at line 28
/Users/peter/Documents/Code.nosync/graph_llm/llm-graph-builder/backend/src/shared/common_fn.py:89: LangChainDeprecationWarning: The class `HuggingFaceEmbeddings` was deprecated in LangChain 0.2.2 and will be removed in 1.0. An updated version of the class exists in the :class:`~langchain-huggingface package and should be used instead. To use it run `pip install -U :class:`~langchain-huggingface` and import as `from :class:`~langchain_huggingface import HuggingFaceEmbeddings``.
  embeddings = SentenceTransformerEmbeddings()

**Comment by kaustubh-darekar** [comment]
can you confirm if you have defined  model config in backend .env as per below format?
LLM_MODEL_CONFIG_openai_gpt_3.5="gpt-3.5-turbo-0125,openai_api_key"
LLM_MODEL_CONFIG_openai_gpt_4o_mini="gpt-4o-mini-2024-07-18,openai_api_key"
LLM_MODEL_CONFIG_openai_gpt_4o="gpt-4o-2024-11-20,openai_api_key"
(whichever is applicable.) (replace openai_api_key with your key)

**Comment by DunaSpice** [comment]
Yes. U I have tried with same result. Let me show you how they look.
LLM_MODEL_CONFIG_openai_gpt_4o=
"gpt-4o-2024-11-20,sk-pro__________MSgO9KPD-9vB6p3oA"

**Comment by DunaSpice** [comment]
From what i understand, i have issues with .env thanks. Will sleep on it.
And get to it tomorrow. Thanks for the help.

**Comment by DunaSpice** [comment]
@kaustubh-darekar @jexp @nielsdejong Thanks. I have understood the logic of the setup. I will try to lay it out here for now. But can take care of it in docs if needed. U need to prepare the env variables manually: (that said I still have to understand the settings for local 2 component(front/back) .env settings for Ollama mode but I have openai working. If you can post a sample config for local deploy with max llms would be amazing! 

that said this is to me more understandable:
LLM_MODEL_CONFIG_openai_gpt_4o_mini = "gpt-4o-mini-2024-07-18,sk-exampleopenaiapikey123456789"
I presumed that in the current example (openai_api_key will be loaded automatically later :)
LLM_MODEL_CONFIG_openai_gpt_4o_mini="gpt-4o-mini-2024-07-18,**openai_api_key**"



also not I am not sure why these are also mentioned then? what is the point of repetition. This also throws off.

OPENAI_API_KEY = "sk-exampleopenaiapikey123456789"
DIFFBOT_API_KEY = "examplediffbotapikey123456"
GROQ_API_KEY = "examplegroqapikey123456"

also having 3 readme files is super confusing. ) in my personal opinion. if you could please help guide me with working trusted development settings, would be great. If I am first to chase, will post here. And thanks again guys.

**Comment by DunaSpice** [comment]
upd. LLM_MODEL_CONFIG_ollama_llama3="llama3.2:latest,http://localhost:11434/"
works charms for llama3.2:latest match to your model in (ollama list) command prompt

so here is experiment one
leave .env in frontend directory set to VITE_LLM_MODELS=""
u will see 13 models. Will see why that is later.
so when in this variable you set
VITE_LLM_MODELS="diffbot,openai_gpt_4o_mini,groq_llama3_70b,ollama_llama3,openai_gpt_4o"
u in essence limit the application. ? not sure yet. and still not sure of groq base api. Will post when have it

**Comment by DunaSpice** [comment]
for the Groq API works with. LLM_MODEL_CONFIG_groq_llama3_70b="llama-3.2-3b-preview,https://api.groq.com/openai/v1:,sk-exampleopenaiapikey123456789"

**Comment by DunaSpice** [comment]
so another finding is that - Model: LLM_MODEL_CONFIG_openai_gpt_4o is a requirement? Used for summaries from what I figure. Maybe worth also changing in the readme

**Comment by kartikpersistent** [comment]
Hi @DunaSpice so finally what modifications do you suggest to add in the readme??

**Comment by SwiftyProjects** [comment]
Hi! @DunaSpice I encountered and resolved this same error "'NoneType' object has no attribute 'split'." I believe the issue involves a specific naming convention difference between frontend model references and backend API configurations.

Ref my issues post: https://github.com/neo4j-labs/llm-graph-builder/issues/910#issue-2708321269

## Root Cause
The error occurs due to two distinct naming requirements:

1. **Backend API Model Names** - Uses hyphens inside quotation marks for actual API calls:

```environment
LLM_MODEL_CONFIG_openai_gpt_4o_mini = "gpt-4o-mini-2024-07-18,your_api_key"
LLM_MODEL_CONFIG_openai_gpt_4 = "gpt-4-turbo-2024-04-09,your_api_key"
```

2. **Frontend Model References** - .env must use underscores for model names inside quotation marks even though suggests hyphens in ~/frontend/example.env , e.g., working syntax:

```environment
VITE_LLM_MODELS="diffbot,openai_gpt_4o_mini,openai_gpt_4,openai_gpt_4o"
```

vs. non-working syntax that produces 'NoneType' warning/failure:

```environment
VITE_LLM_MODELS="diffbot,openai-gpt-3.5,openai-gpt-4o"
```

---

## Altogether, Working Configuration Examples

### Backend (.env)

```environment
# Variable names use underscores
# API model names (in quotes) use hyphens as per OpenAI's API requirements
LLM_MODEL_CONFIG_openai_gpt_4o_mini = "gpt-4o-mini-2024-07-18,your_full_api_key"
LLM_MODEL_CONFIG_openai_gpt_4 = "gpt-4-turbo-2024-04-09,your_full_api_key"
LLM_MODEL_CONFIG_openai_gpt_4o = "gpt-4o-2024-08-06,your_full_api_key"
LLM_MODEL_CONFIG_diffbot = "diffbot,your_full_diffbot_key"
```

### Frontend (.env)
```environment
# Both variable names AND model references must use underscores
VITE_LLM_MODELS="diffbot,openai_gpt_4o_mini,openai_gpt_4,openai_gpt_4o"
VITE_ENV="DEV"
VITE_LLM_MODELS_PROD="diffbot,openai_gpt_4o_mini,openai_gpt_4,openai_gpt_4o"
```


## Key Points
1. Backend config maintains two formats:
   - Variable names use underscores: `LLM_MODEL_CONFIG_openai_gpt_4o_mini`
   - API model names (in quotes) use hyphens: `"gpt-4o-mini-2024-07-18"`
2. Frontend config must use underscores throughout
3. The current example.env in frontend may be misleading as it suggests using hyphens

---

## Note on Model Selection

- *note: I've included diffbot too, just as example with no hyphens nor underscore for diffbot, so should work either way as non-issue

- **double note: While gpt-3.5-turbo works fine when following the above naming convention fix, I recommend supplanting 3.5 with gpt-4o-mini instead. This is because OpenAI API states: "As of July 2024, gpt-4o-mini should be used in place of gpt-3.5-turbo, as it is cheaper, more capable, multimodal, and just as fast. gpt-3.5-turbo is still available for use in the API." <https://platform.openai.com/docs/models>

Hope this helps clarify the naming requirements and model selection!

Cheers!

**Comment by jexp** [comment]
Fixed in documentation
